socks/pool.go

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2019 The Decred developers
+// Use of this source code is governed by a 3-clause BSD
+// license that can be found in the LICENSE file.
+
+package socks
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+)
+
+type Pool struct {
+	mtx     sync.Mutex
+	numOpen uint32
+	maxOpen uint32
+
+	proxy Proxy
+}
+
+func NewPool(proxy Proxy, maxOpen uint32) *Pool {
+	pool := Pool{
+		maxOpen: maxOpen,
+		proxy:   proxy,
+	}
+
+	return &pool
+}
+
+func (p *Pool) DialContext(ctx context.Context, network, address string) (net.Conn, error) {
+	select {
+	case <-ctx.Done():
```

Comment:
If you don't need to use a multi-case select, it's a little shorter and more efficient to instead write:

```
if err := ctx.Err(); err != nil {
    return err
}
```